components/script/dom/htmlelement.rs

```diff
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
         DOMString::from(text)
     }
 
-    // https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute
+    /// <https://html.spec.whatwg.org/multipage/#set-the-inner-text-steps>
     fn SetInnerText(&self, input: DOMString) {
+        // Step 1: Let fragment be the rendered text fragment for value given element's node
+        // document.
+        let fragment = self.rendered_text_fragment(input);
+
+        // Step 2: Replace all with fragment within element.
+        Node::replace_all(Some(fragment.upcast()), self.upcast::<Node>());
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-outertext>
+    fn GetOuterText(&self) -> Fallible<DOMString> {
+        let node = self.upcast::<Node>();
+        let window = window_from_node(node);
+        let element = self.as_element();
+
         // Step 1.
+        let element_not_rendered = !node.is_connected() || !element.has_css_layout_box();
+        if element_not_rendered {
+            return Ok(node.GetTextContent().unwrap());
+        }
+
+        window.layout_reflow(QueryMsg::ElementOuterTextQuery);
+        let text = window
+            .layout()
+            .query_element_outer_text(node.to_trusted_node_address());
+
+        Ok(DOMString::from(text))
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute%3Adom-outertext-2>
```

Comment:
https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute:dom-outertext-2